x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -0,0 +1,349 @@
+            if (sameUsername && ChangePasswordAction.NAME.equals(action)) {
+                return checkChangePasswordAction(authentication);
+            }
+
+            assert AuthenticateAction.NAME.equals(action) || HasPrivilegesAction.NAME.equals(action)
+                || GetUserPrivilegesAction.NAME.equals(action) || sameUsername == false
+                : "Action '" + action + "' should not be possible when sameUsername=" + sameUsername;
+            return sameUsername;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean shouldAuthorizeIndexActionNameOnly(String action, TransportRequest request) {
+        switch (action) {
+            case BulkAction.NAME:
+            case IndexAction.NAME:
+            case DeleteAction.NAME:
+            case INDEX_SUB_REQUEST_PRIMARY:
+            case INDEX_SUB_REQUEST_REPLICA:
+            case DELETE_SUB_REQUEST_PRIMARY:
+            case DELETE_SUB_REQUEST_REPLICA:
+            case MultiGetAction.NAME:
+            case MultiTermVectorsAction.NAME:
+            case MultiSearchAction.NAME:
+            case "indices:data/read/mpercolate":
+            case "indices:data/read/msearch/template":
+            case "indices:data/read/search/template":
+            case "indices:data/write/reindex":
+            case "indices:data/read/sql":
+            case "indices:data/read/sql/translate":
+                if (request instanceof CompositeIndicesRequest == false) {
+                    throw new IllegalStateException("Composite and bulk actions must implement " +
+                        CompositeIndicesRequest.class.getSimpleName() + ", " + request.getClass().getSimpleName() + " doesn't");
+                }
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    @Override
+    public void authorizeIndexActionName(Authentication authentication, TransportRequest request, String action,
+                                         AuthorizationInfo authorizationInfo, ActionListener<AuthorizationResult> listener) {
+        if (authorizationInfo instanceof RBACAuthorizationInfo) {
+            final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
+            if (role.indices().check(action)) {
+                listener.onResponse(AuthorizationResult.granted());
+            } else {
+                listener.onResponse(AuthorizationResult.deny());
+            }
+        } else {
+            listener.onFailure(new IllegalArgumentException("unsupported authorization info:" +
+                authorizationInfo.getClass().getSimpleName()));
+        }
+    }
+
+    @Override
+    public List<String> loadAuthorizedIndices(Authentication authentication, TransportRequest request, String action,
+                                              AuthorizationInfo authorizationInfo, Map<String, AliasOrIndex> aliasAndIndexLookup) {
```

Comment:
Can we not have the `TransportRequest` as a parameter here? I know it is omnipresent, but having it here is confusing because the method returns a super set of the indices irrespective of the request. Another implementer might be confused: should I check the indices in the request and return only the authorized ones or should there be all authorized indices for this action.

Reply:
I see what you are saying and this makes sense